### DISCUSSION TITLE
How to reference/use customize icons in omnet++ sim?

### DISCUSSION THREAD

**Discussion by jeromesolomon** [omnetpp/omnetpp#1135] [General]
Dear Reader,

Thank you in advance for any guidance/suggestions.

I'm relatively new to veins. I'd like to use custom .png icons for my simulation. I'm able to change the icon by referencing it as such in my Scenario.ned file. But, the icon is missing when the sim runs.

I reference the custom icon as follows:

```
rsu[1]: RSU {
    @display("p=2000,2000;i=veins/avswu_icons/rsu;is=l");
};
```

the file is `/home/veins/src/veins/images/veins/avswu_icons/rsu_l.png`

I can see these icons appear when I open the .ned file in eclipse.  But, when the simulation runs these icons are not used. 

**Is there a way to use your own custom icons in omnet++? If so, he would I reference the custom icons for the simulation itself?**

I am working in the instant veins virtualbox under debian.

See images below. red arrows show the missing icons.

### .ned file shows correct custom icons (green arrows)
![correct-icons](https://github.com/omnetpp/omnetpp/assets/11489355/08db90a6-7828-411a-ba7c-f00639925557)

### Veins sim running with missing icons (red arrows)
![missing-icons](https://github.com/omnetpp/omnetpp/assets/11489355/02e92c8c-43e1-48ff-8e2c-833eda9bb6a5)

**Comment by rhornig** [comment]
Usually, the 'images' folder that contains the project specific images are placed in the ROOT of the project folder. You seemingly placed it in `ROOT/src/veins`. The run scripts that start the simulation is also written usually to assume the location in the ROOT/images folder. (the fact that the icon is visible in the IDE is just pure coincidence).

**Reply by jeromesolomon** [reply]
Dear @rhornig, et al, and friends :)

First! :) Thank you for your reply.  Based on your guidance, I also found the [documentation under 8.4.9.1 The Image Path](https://doc.omnetpp.org/omnetpp/manual/) explains exactly what you described. But, it still does not change the icon in the simulation for me (under linux, instant-veins, omnet, veins 5.7.x) as explained above. 

I also tried to change the image-path in omnet*.ini and it seemingly had no effect.

For another test, I painted pink on the bicycle_l.png image itself within the images/veins/node directory and the simulation defaults back to the regular bicycle when it runs. The simulation images and the .ned IDE images appear to be totally disconnected (not referenced by path nor other means).

Can you explain how this is coincidence? I may be misunderstanding somehow. Apologies.

Were you able to change one of your icons to a custom image and see it during simulation? Are you using linux/instant-veins/omnet?

-Gabriel

**In Omnet++ IDE**
![bicycle-pink](https://github.com/omnetpp/omnetpp/assets/11489355/854c805c-8323-4774-ab27-324d657b2722)

**During the simulation**
![bicycle-normal](https://github.com/omnetpp/omnetpp/assets/11489355/eb90f477-88ce-4c01-a7b8-22658f79c861)

**Reply by rhornig** [reply]
I'm not sure what is the command line used to launch the simulation. Check the scripts that are used to start the simulation because image path parameters can be specified also on command line and command line options take precedence over values specified in INI file.

A brief look into the veins_run command shows that it definitely uses the --image-path command line option.

**Reply by jeromesolomon** [reply]
Thanks so much for trying to help @rhornig !

If someone has a working example of using custom images/icons in their simulation when it runs, and can post it, that would be great.

In appreciation.

**Reply by rhornig** [reply]
INET has tons of custom icons.

**Reply by jeromesolomon** [reply]
Thank you, and apologies. By "custom", I mean user created as explained herein. They are 
 for our research and we wish to create them by hand.

**Reply by rhornig** [reply]
Well. INET Framework is technically a "user" library using OMNeT++. But if you want to go deeper into the rabbit hold, Simu5G also has custom icons: https://github.com/Unipisa/Simu5G/

**Comment by jeromesolomon** [comment]
I like to consider myself a volant computer scientist, so I do not do rabbit holes. :)  Yet, if there is a reproduceable isolated test case for setting image path and it using a custom icon during sim, then I can move forward on this.  Moreover, all is well. This is a nice to have. As you said, there are a lot of icons -- just not any that represent the roles of nodes in our research. Thank you for trying to help out.  Take care, and have a good week sir!

**Reply by rhornig** [reply]
well. You can look also into the samples/petrinet example which also uses custom images (i.e. images inside the project folder and not in the default OMNeT++ image folder).

If you want to troubleshoot an image loading issue, the first thing is to take a look at the console output of the simulation when you start it. It prints out the IMAGE folders and the number of images found in it so you can cross check whether it accesses the folder

**Reply by jeromesolomon** [reply]
Thanks @rhornig . Those are good examples, that should work! But, alas it's nor working for me under 5.7 nor is canvas/samples method. I even tried set programmatically using setTag() but it has the same behavior. Custom images appear in .ned, but when the sim executes the image is missing.

**Comment by jeromesolomon** [comment]
Good news. I was able to get it to work using this in my code.

`findHost()->getDisplayString().setTagArg("i", 0, "veins/your_folder/your_custom_icon");`

Thanks for your pointers @rhornig
